node/src/service.rs

```diff
@@ -214,13 +220,66 @@ pub fn new_light(config: Configuration) -> Result<impl AbstractService, ServiceE
             );
             Ok(import_queue)
         })?
-        .build()
+        .build()?;
+
+    register_difficulty_metric(&service, &pow_alg)?;
+    Ok(service)
+}
+
+fn register_difficulty_metric<S>(service: &S, pow_alg_config: &PowAlgConfig) -> Result<(), Error>
+where
+    S: AbstractService,
+    sc_client::Client<S::Backend, S::CallExecutor, S::Block, S::RuntimeApi>: AuxStore,
+{
+    match pow_alg_config {
+        PowAlgConfig::Dummy => register_difficulty_metric_impl::<DummyPowDifficulty, _>(service),
+        PowAlgConfig::Blake3 => register_difficulty_metric_impl::<Blake3PowDifficulty, _>(service),
+    }
+}
+
+fn register_difficulty_metric_impl<Difficulty, S>(service: &S) -> Result<(), Error>
+where
+    S: AbstractService,
+    sc_client::Client<S::Backend, S::CallExecutor, S::Block, S::RuntimeApi>: AuxStore,
+    Difficulty: parity_scale_codec::Decode + Default + TryInto<u64> + Send + 'static,
+{
+    use futures::future::ready;
+    use substrate_prometheus_endpoint::{Gauge, U64};
+
+    let registry = match service.prometheus_registry() {
+        Some(registry) => registry,
+        None => return Ok(()),
+    };
+    let unregistered_gauge = Gauge::<U64>::new(
+        "best_block_difficulty",
+        "The difficulty of the best block in the chain",
+    )
+    .map_err(|error| format!("failed to create a metric gauge: {}", error))?;
+    let gauge = substrate_prometheus_endpoint::register(unregistered_gauge, &registry)
+        .map_err(|error| format!("failed to register a metric gauge: {}", error))?;
+    let client = service.client();
+    let metric_stream = client
+        .import_notification_stream()
+        .filter(|info| ready(info.is_new_best))
+        .filter_map(move |info| {
+            ready(
+                sc_consensus_pow::PowAux::<Difficulty>::read::<_, S::Block>(&*client, &info.hash)
+                    .ok(),
+            )
+        })
+        .map(|difficulty| difficulty.difficulty.try_into().unwrap_or(u64::MAX))
+        .for_each(move |difficulty| {
+            gauge.set(difficulty);
+            ready(())
+        });
```

Comment:
We’re using a lot of stream methods. I think it would be more readable, less complicated, and more performant to just call `for_each` on and have everything in the handler.
```suggestion
        .map(|difficulty| difficulty.difficulty.try_into().unwrap_or(u64::MAX))
        .for_each(move |info| {
            if info.is_new_best {
                let pow_aux_result = sc_consensus_pow::PowAux::<Difficulty>::read::<_, S::Block>(&*client, &info.hash);
                match pow_aux_result {
                    Ok(pow_aux) =>  {
                        let difficulty = pow_aux.difficulty.try_into().unwrap_or(u64::MAX);
                        gauge.set(difficulty);
                    },
                    Error(err) => todo!("log error"),
                }
            }
            ready(())
        });
```
Maybe it also makes sense to extract the `PowAux::read` and `.difficulty.try_into()` code into a separate function.

Reply:
I've turned it into a single closure. It is a bit more readable and easier to understand. I wouldn't mind performance, the compiler is really good in inlining closures and removing unnecessary steps.

I haven't extracted `PowAux::read` and `.difficulty.try_into()` to a separate function, because its signature contained some very complex generics and it would only replace 2 rather clear oneliners.